src/main/java/io/github/bookcrawler/core/impl/publio/PublioBookLinksCrawler.java

```diff
@@ -0,0 +1,82 @@
+package io.github.bookcrawler.core.impl.publio;
+
+import io.github.bookcrawler.core.BooksLinkCrawler;
+import io.github.bookcrawler.core.SourceScrapper;
+import io.github.bookcrawler.core.impl.SourceScrappingResult;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.github.bookcrawler.core.BookStore.PUBLIO;
+
+public class PublioBookLinksCrawler implements BooksLinkCrawler {
+    @Override
+    public List<String> crawl(String url, SourceScrapper sourceScrapper) {
+
+        SourceScrappingResult sourceScrappingResult = sourceScrapper.scrap(url);
+        if (!sourceScrappingResult.isSuccessful()) {
+            return Collections.emptyList();
+        }
+
+        return getPageUrls(url, sourceScrapper, sourceScrappingResult)
+                .stream()
+                .map(sourceScrapper::scrap)
+                .filter(SourceScrappingResult::isSuccessful)
+                .flatMap(e -> getBookElements(e.getSource()).stream())
+                .map(e -> getLink(href(e)))
+                .collect(Collectors.toList());
+    }
+
+    private ArrayList<String> getPageUrls(String url, SourceScrapper sourceScrapper, SourceScrappingResult sourceScrappingResult) {
+        Document source = sourceScrappingResult.getSource();
+        Integer pages = getPageQuantity(source);
+
+        ArrayList<String> pageUrls = new ArrayList<>();
+        pageUrls.add(url);
+        for (int i = 1; i < pages; i++) {
```

Comment:
How about a stream instead? One operational λ and filtering on scrapping being successful or not and you're done.